src/main/java/org/jboss/logmanager/LogContextRouter.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ *
+ * Copyright 2019 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.logmanager;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A router used to route log messages to a specific log context based on the current threads
+ * {@linkplain Thread#getContextClassLoader() context class loader}.
+ *
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ */
+@SuppressWarnings("unused")
+public class LogContextRouter {
```

Comment:
Making a note here so I don't forget my though process. Maybe this becomes an interface which is set on the `LogContext` which WildFly implements and by default there is no implementation. The `LoggerNode`? would check if the context has a router, then invoke something like `LogContextRouter.publish(record)`.